core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java

```diff
@@ -0,0 +1,230 @@
+import org.infinispan.commons.CacheException;
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.container.DataContainer;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.InboundTransferTask;
+import org.infinispan.statetransfer.StateChunk;
+import org.infinispan.topology.CacheTopology;
+
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * @author Ryan Emerson
+ * @since 9.0
+ */
+public class StateReceiverImpl<K, V> implements StateReceiver<K, V> {
+
+   private static Log log = LogFactory.getLog(StateReceiverImpl.class);
+   private static boolean trace = log.isTraceEnabled();
+
+   private String cacheName;
+   private CommandsFactory commandsFactory;
+   private DataContainer<K, V> dataContainer;
+   private RpcManager rpcManager;
+   private long transferTimeout;
+
+   private final Object lock = new Object();
+
+   @GuardedBy("lock")
+   private final Map<K, Map<Address, InternalCacheEntry<K, V>>> keyReplicaMap = new HashMap<>();
+
+   @GuardedBy("lock")
+   private final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+
+   private CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> completableFuture = CompletableFuture.completedFuture(null);
+
+   private CacheTopology cacheTopology;
+   private int segmentId;
+   private int minTopologyId = -1;
+
+   @Inject
+   public void init(Cache cache,
+                    CommandsFactory commandsFactory,
+                    Configuration configuration,
+                    DataContainer<K, V> dataContainer,
+                    RpcManager rpcManager) {
+      this.cacheName = cache.getName();
+      this.commandsFactory = commandsFactory;
+      this.dataContainer = dataContainer;
+      this.rpcManager = rpcManager;
+
+      this.transferTimeout = configuration.clustering().stateTransfer().timeout();
+   }
+
+   @Override
+   public boolean isTransferInProgress() {
+      synchronized (lock) {
+         return !completableFuture.isDone();
+      }
+   }
+
+   @Override
+   public CompletableFuture<List<Map<Address, InternalCacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId) {
+      synchronized (lock) {
+         // TODO improve handling of concurrent invocations
+         if (!completableFuture.isDone()) {
+            if (trace) log.tracef("Concurrent invocations of getAllReplicasForSegment detected %s", completableFuture);
+            throw new IllegalStateException("It is not possible to request multiple segments of a cache simultaneously.");
+         }
+
+         this.segmentId = segmentId;
+         ConsistentHash hash = cacheTopology.getReadConsistentHash();
```

Comment:
Ok, I've updated the code so that `getWriteConsistentHash()` is always used as it is more explicit than `getConsistentHash()`